src/util/address.rs

```diff
@@ -727,6 +727,44 @@ impl Address {
             (Network::Testnet, _) | (Network::Regtest, _) | (Network::Signet, _) => true
         }
     }
+
+    /// Returns true if the given pubkey is directly related to the address payload.
+    /// This is determined by directly comparing the address payload with either the
+    /// hash of the given public key or the segwit redeem hash generated from the
+    /// given key. For taproot addresses, the supplied key is assumed to be tweaked
+    pub fn is_related_to_pubkey(&self, pubkey: &PublicKey) -> bool {
+        match self.address_type() {
+            Some(AddressType::P2pkh) | Some(AddressType::P2sh) | Some(AddressType::P2wpkh) | Some(AddressType::P2wsh) => {
+                let pubkey_hash = pubkey.pubkey_hash();
+                let payload = self.payload_as_bytes();
+
+                (*pubkey_hash == *payload) || (*segwit_redeem_hash(&pubkey_hash) == *payload)
+            },
+            Some(AddressType::P2tr) => {
+                let xonly_pubkey = XOnlyPublicKey::from(pubkey.inner);
+                self.is_related_to_xonly_pubkey(&xonly_pubkey)
+            }
+            None => false
+        }
+    }
+
+    /// Returns true if the supplied xonly public key can be used to derive the
+    /// address. This will only work for Taproot addresses. The Public Key is
+    /// assumed to have already been tweaked.
+    pub fn is_related_to_xonly_pubkey(&self, xonly_pubkey: &XOnlyPublicKey) -> bool {
+        let tweaked_pubkey = TweakedPublicKey::dangerous_assume_tweaked(*xonly_pubkey);
+        let derived_address = Address::p2tr_tweaked(tweaked_pubkey, self.network);
+        *self == derived_address
```

Comment:
Any particular reason why you didn't try to extract the key from address and compare that? (I don't see a problem, just curious as it seems less natural to me.)

Reply:
That makes much more sense and removes the need for most of the lines that I added for TR support. That is a much bette ridea. Just pushed an updated